### ISSUE TITLE
[Question] Cannot reproduce GPy with GPyTorch for 3D problem

### ISSUE THREAD

**Issue by elindgren** [cornellius-gp/gpytorch#1757]
Hi!

I'm currently working on porting a simple GPy model to GPyTorch. It's for a regression task, in which the input data (X) are coordinates in the range X \in [0, 7] and the targets are values in the range [-158, 3067]. At the moment I'm not interested in extrapolation, only interpolation. 

The GPy model I'm using is defined as follows:

```python
def build_model(X_data: np.ndarray, y_data:np.ndarray, domain: Domain, mode: int):
    """Return GP model given a set of data and a domain.
    """
    if mode == 1:
        k1 = GPy.kern.RBF(input_dim=len(domain))
        k2 = GPy.kern.Bias(input_dim=len(domain))
        kernel = k1 + k2
        kernel['rbf.lengthscale'].set_prior(GPy.priors.Gamma(a=1, b=2))
    else:
        raise ValueError(f'Unknown mode: {mode}')
    model = GPy.models.GPRegression(X_data, y_data, kernel, normalizer=GPy.util.normalizer.Standardize())
    model.optimize()
    _ = model.optimize_restarts(verbose=True)
    return model
```

This yields the following model
![gpy_model](https://user-images.githubusercontent.com/27208255/133790031-33a48160-e636-4795-8fc5-ed789dd9b721.png)

and the predicted function values in the range [2.5,7] (since the function grows rapidly as it approaches zero) like this:
![gpy_pes](https://user-images.githubusercontent.com/27208255/133788078-f606edf5-1938-4582-b9d0-871ecb4bd478.png)

For the GPyTorch model, I've followed the ExactGP regression tutorial. I've modified the model slightly to perform z-scaling of the input values, due to large range of my y values (I do it for the X-positions as well, but that doesn't seem to have an effect). Here is my model

```python
import numpy as np
import torch
import gpytorch
import matplotlib.pyplot as plt

class SimpleGPyTorch(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood, mode, standardize):
        super(SimpleGPyTorch, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        # Single RBF kernel, no bias kernel
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel())
        # self.covar_module = gpytorch.kernels.SpectralMixtureKernel(num_mixtures=4, ard_num_dims=3)
        self.mode = mode # either 1 for CPU or 2 for GPU
        self.standardize = standardize
        if standardize:
            # Save mean/std of input for standardization
            self.y_mean = train_y.mean()
            self.y_std = train_y.std()
            self.x_mean = train_x.mean(dim=0, keepdim=True)
            self.x_std = train_x.std(dim=0, keepdim=True)
            self.standardize_training_data()
        

    def forward(self, x):
        mean = self.mean_module(x)
        covar = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean, covar)


    def optimize(self, training_iter=125, verbose=False, plot=True):
        model = self
        likelihood = self.likelihood
        X = self.train_inputs[0]
        y = self.train_targets

        model.train()
        likelihood.train()
        optimizer = torch.optim.Adam(model.parameters(), lr=0.1)
        # optimizer = FullBatchLBFGS(model.parameters(), lr=0.01)

        mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
        loss_trace = []
        
        def closure():
            optimizer.zero_grad()
            output = model(X)
            loss = -mll(output, y) # reach MLE through gradient descent
            return loss
        with gpytorch.settings.cg_tolerance(0.01), gpytorch.settings.cg_tolerance(10000), gpytorch.settings.max_preconditioner_size(100):
            for i in range(training_iter):
                # Set gradients from previous iteration to 0
                loss = closure()
                loss.backward()
                optimizer.step()

                # options = {'closure': closure, 'current_loss': loss, 'max_ls': 20}
                # loss, _, _, _, _, _, _, fail = optimizer.step(options)
                if verbose and i%50 == 0:
                    print(f'Iteration {i} - Loss: {loss.item():.3f} - Lengthscale: {model.covar_module.base_kernel.lengthscale.item():.3f} - Noise: {model.likelihood.noise.item():.3f}')
                loss_trace.append(loss.detach().numpy())

                # if fail: 
                #     break
            if plot:
                _, ax = plt.subplots(figsize=(8,6))
                ax.set_xlabel("Training iteration")
                ax.set_ylabel("Marginal Log Likelihood Loss")
                ax.plot(loss_trace)
        return loss_trace


    def standardize_training_data(self):
        # Standardize targets
        self.train_targets -= self.y_mean
        self.train_targets /= self.y_std
        # Standardize features
        train_x = self.train_inputs[0]
        train_x -= self.x_mean
        train_x /= self.x_std
        self.train_inputs = (train_x,)


    def predict(self, x):
        self.eval()
        self.likelihood.eval()
                
        with torch.no_grad(), gpytorch.settings.fast_pred_var(): 
            x_pred = torch.from_numpy(x).type(torch.FloatTensor)

            if self.standardize:
                # Standardize prediction features
                x_pred = (x_pred - self.x_mean) / self.x_std
                
        
            prediction = self.likelihood(self(x_pred))
            mean = prediction.mean.detach().numpy()
            var = prediction.variance.detach().numpy()

            if self.standardize:
                # Rescale prediction to original training data scale
                original_mean = self.y_mean.detach().numpy()
                original_std = self.y_std.detach().numpy()
                mean = mean*original_std + original_mean
                var = var*original_std**2 # Variance is stationary and is only changed by a factor - https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/gaussian_process/_gpr.py#L355
            return mean, var
```

which yields the following training curve and model parameters.
![gpytorch_mll](https://user-images.githubusercontent.com/27208255/133789881-0f31d2f7-8da8-474c-ab2b-b5b6ff2e64b9.png)
![gpytorch_params](https://user-images.githubusercontent.com/27208255/133789944-7be69ee0-0a0e-4d71-b867-2cb8a33362b3.png)

and the predicted y values:
![gpytorch](https://user-images.githubusercontent.com/27208255/133789868-b84fb15f-fd99-4f4d-9e90-05c8feab6d55.png)

I've tried different optimizers (Adam, SGD, LBFGS from PyTorch as well as [PyTorch-LBFGS](https://github.com/hjmshi/PyTorch-LBFGS)), learning rates, using a gamma prior on the lengthscale, changing the kernel (to the GaussianMixture kernel), increasing the accuracy of the CG solves and the preconditioner size. I've also tried restarting training to avoid getting stuck in local minima, but in those case I've found that the model sometimes converges to predicting a flat function. 

My guess is that the poor performance of the GPyTorch model is due to the model not converging very well during training. I see much better results when capping the range of y values to for instance [0, 50] instead of the full range, but this is not necessary for GPy. 

As a sanity check, I've tested these exact models for both a toy problem (a noisy sine) as well as a simpler version of the problem above (2D positions instead of 3D), in which both GPy and GPyTorch agrees almost perfectly. For the 3D problem, I've also tried implementing the same model using scikit-learn, and there I get very similar results as for GPy. 
Is there something I'm missing?

**Comment by Balandat** [comment]
what's the size of the data set here?

**Comment by elindgren** [comment]
> what's the size of the data set here?

Hi! Thanks for the reply!

I have ~850 data points in this case.

**Comment by Balandat** [comment]
This is over the 800 threshold at which gpytorch will use LinearCG/lanczos for computations. Might be a red herring, but what happens if you use fewer (say 500) datapoints, do these discrepancies still exist? The other thing I would try is to run this in double dtype.

**Comment by elindgren** [comment]
@Balandat With 500 datapoints and double precision the convergence is much smoother, and the predicted values look closer to both GPy and Scikit. However, there are still some discrepancies, both in the overall shape of the predicted function as well as the standard deviation. Interestingly, the model doesn't seem very well converged for 500 datapoints; I get a very different shape of the predicted function between various runs.

To test this idea of convergence, I increased the number of training iterations, which I wasn't able to earlier due to numerical instabilities. Here is a plot of the predicted function for 500 datapoints
![500](https://user-images.githubusercontent.com/27208255/133973323-1a004766-9c30-4e85-94bf-da0e6f46fed3.png)

And for 750 datapoints
![750](https://user-images.githubusercontent.com/27208255/133973255-b4ffb893-0a53-44e6-959c-d912b93588b1.png)


However, when I try to run this with multiple restarts I almost always end up with a very flat predicted function as the lowest loss model. This very rarely happens when I train a single model. Does this have something to do with the internal state of GPyTorch (random seeds etc.)? Here is for instance a plot of the predicted function with 10 restarts for 750 datapoints, which reaches a similar loss, lengthscale and noise as the non-flat plot for 750 datapoints above.

![flat750](https://user-images.githubusercontent.com/27208255/133974014-eaff89ea-5c25-403a-af30-01b09a3e18d4.png)

For reference, here are the converged values for 750 datapoints
**Flat**
- Lengthscale: 0.70324
- Noise: 0.0001145
- Loss: -1.34402

**Non-flat**
- Lengthscale: 0.69108
- Noise: 0.0001215
- Loss: -1.178

**Comment by elindgren** [comment]
Here is my code for performing multiple restarts, to compare against:

```python 
# Restart training to avoid getting stuck in local minima
    score = []
    models = []
    for i in range(10):
        # Initialize likelihood and model
        likelihood = gpytorch.likelihoods.GaussianLikelihood()
        model = SimpleGPyTorch(X, y, likelihood, mode, standardize)
        model = model.double()

        loss = model.optimize(training_iter=500, plot=True)
        best_loss = min(loss)
        print(best_loss)

        score.append(best_loss)
        models.append(model)

    # Find best model
    score = np.array(score)
    best_index = np.argmin(score, axis=0)   
    return models[best_index]
```

**Comment by Balandat** [comment]
So there could be two issues here:
- the MLL has multiple local optima close together and you're landing in different ones depending on the initialization
- the MLL has a single (or an obvious) global maximum and is quite peaked, and you're having a hard time finding that maximum using Adam. If the computation doesn't use approximate methods and things are deterministic, I've found a quasi-second order optimizer such as L-BFGS to work quite well. But LBFGS from PyTorch is crap, and PyTorch-LBFGS is a stochastic quasi-second order optimizer and may not work as well for that reason. 

You could try using the [`fit_gpytorch_model`](https://github.com/pytorch/botorch/blob/main/botorch/fit.py#L34) utility we have in BoTorch and see how well that works for you: 
```
from botorch.fit import fit_gpytorch_model

mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
fit_gpytorch_model(mll)
```

As to the flat result when using random restarts: I am not sure what's going on there, I didn't see any obvious errors in the code.

**Comment by elindgren** [comment]
Using fit_gpytorch_model from BoTorch lead to a better optimized model (loss=-3.4 compared to -1.38 for my Adam implementation). However, when computing the RMSE the gpytorch implementation is still behind GPy; for GPyTorch I obtain an RMSE of ~530, whilst for GPy I obtain an RMSE of 0.811. For a toy example of a noisy sine that I've implemented as a sanity check I get exactly the same RMSE for GPy and GPyTorch (with or without BoTorch). The problems thus seems to be related to this 3D setting. 

![botorch](https://user-images.githubusercontent.com/27208255/134165448-96f71ca0-1706-48b4-8cb5-98e3e5eae4dc.png)

The optimization is from what I can tell fairly flat (barring that I might not have sampled the parameterspace well enough). In the figure below I've calculated the marginal likelihood for my data using GPy when fixing RBF lengthscale and variance (outputscale in GPyTorch if I've understood correctly), after which I've optimized the remaining hyperparameters (Bias kernel mean as well as noise). Plotted are also the true maximum of the likelihood, i.e. the highest likelihood value sampled, as well as the position of GPyTorch and GPy otimized models. Note that the model I obtain using Adam in GPyTorch is plotted as a red dot, whilst the one I obtain with GPyTorch & BoTorch is plotted as a red star. I think it's quite weird that Adam performs poorly in this setting; it shouldn't get stuck in "flat" regions due to the adaptive learning rate right? 

![convergence_landscape](https://user-images.githubusercontent.com/27208255/134167782-12ab94e0-8b1a-4fd5-91ad-101f7e8a775c.png)


Furthermore, I've also tried setting the same values for lengthscale and outputscale in GPyTorch as I obtain with GPy (by setting really narrow priors in GPyTorch), but that yielded a very large loss, which seems to be in contrast to the optimization landscape seeming to be so flat.

From what I could tell, the BoTorch utility uses scipy.minimize internally with the L-BFGS-B optimizer.  Switching from the scipy optimizer to the included fit_gpytorch_torch yielded the same results as my own Adam optimizer (not surprising since they are basically the same thing from what it seems). Are there any reasons as to why scipy's optimizer is so much better than torch's Adam, in the light of the optimization landscape seeming to be so flat?

**Comment by wjmaddox** [comment]
Yes, this is the difference between first order and second order optimizers generically. First order optimizers like Adam are "faster per iteration" but second order optimizers often require fewer optimization iterations. 

One thing that I'm noticing now is that your GPy model defines a `Gamma(1,2)` prior over the outputscale, but your gpytorch model is not doing that. Maybe you should try including that in your gpytorch model using our prior definitions, as that will certainly bias the model fits and comparisons.

**Comment by jacobrgardner** [comment]
In addition to the prior definition, could you also try changing the default likelihood noise lower bound? I notice in your listed hyperparameters above that the noise is getting quite small, and close to the default lower bound.

When you make your `GaussianLikelihood` use this:
```python
likelihood = gpytorch.likelihoods.GaussianLikelihood(noise_constraint=gpytorch.constraints.Positive())
```

**Comment by elindgren** [comment]
@wjmaddox Alright, so it's not surprising then that Adam gets stuck. But I find it quite odd that L-BFGS-B in BoTorch doesn't converge to the same minima as GPy (which also uses L-BFGS I believe, but don't quote me on that).

Not including the Gamma(1,2) prior over the RBF lengthscale in PyTorch was something I was experimenting with back and forth a few days ago, but during the last few days I've used the prior in both GPy and GPyTorch. I've noticed it makes little to no difference though, since the prior is fairly broad over the range of the lengthscale that I obtain (0.8 for Adam and 1.05 for BoTorch, 1.7 for GPy).

**Comment by elindgren** [comment]
@jacobrgardner Changing the constraint for the likelihood per your suggestion leads to the model converging to a lower loss (mll=-7.3), but with a smaller RBF lengthscale and a very small loss.

![low_bound](https://user-images.githubusercontent.com/27208255/134170441-22ae7b45-4e14-4bbd-bf7d-a13b91380c28.png)

This also leads to a fairly weird looking predicted function. 
![gpytorch_lower_constraint](https://user-images.githubusercontent.com/27208255/134170449-98f3f9e4-8536-4d1b-9d45-29a76f3d8bed.png)

Doing this I also get some warnings from BoTorch (from the L-BFGS-B optimization I believe): "NumericalWarning:A not p.d., added jitter of 1.0e-08 to the diagonal". Does this mean that the optimization hasn't converged very well due to numerical problems?

**Comment by jacobrgardner** [comment]
@elindgren the lower MLL could be expected. GPy technically uses a noise lower bound of 1e-6 in the form of adding jitter when computing cholesky:
https://github.com/SheffieldML/GPy/blob/3e19a85575687e37fd6f61174115d7c94d2c96e6/GPy/util/linalg.py#L65

Removing the lower bound in GPyTorch entirely will let you get potentially much smaller noises than 1e-6 (and therefore potentially lower losses), though at the potential cost of numerical instability as the conditioning of the kernel matrix gets worse. The NumericalWarnings with jitter of 1e-8 don't seem too bad though: 1e-8 is still a really small amount of jitter.

If you want the warnings to go away, you could probably run everything in fp64 instead of torch's default fp32 by adding `torch.set_default_dtype(torch.float64)` at the top of your script.

Alternatively, if you want to set the same lower bound, you can use a `GreaterThan(1e-6)` constraint.

It's not surprising that the lengthscale gets smaller as the noise decreases: smaller lengthscales cause the GP fit to interpolate the training data more tightly, which makes sense in a lower noise regime.

**Comment by elindgren** [comment]
I think I managed to get it working. Using botorch in combination with setting a positive constraint on the noise and increasing the max cholesky size to be larger than the size of my dataset leads to very similar results to GPy. The hyperparameters are different than for GPy, but the predicted model is very similar:

![z-benzene](https://user-images.githubusercontent.com/27208255/134314482-62cee2fd-7b61-41e5-8c0b-09a00499e932.png)

I had also made an error when calculating the MAE and RMSE earlier, so they were closer before than what they seem. Now GPy and GPyTorch are almost identical. 

![mae-gpytorch](https://user-images.githubusercontent.com/27208255/134314332-18a49dc8-0fd5-444d-88ce-ec7eeef6cfc5.png)

The problem with the model predicting a flat function was also an error that I had made: I had forgotten to copy my input tensors to the model, and since I perform z-scaling of both input data and targets on model creation meant that they both went to zero after a few models had been created (since the same tensors were erroneously used for all models). 

@Balandat, @wjmaddox and @jacobrgardner, thank you for taking the time to helping me investigate my problem! And thank you for a very nice and well documented gaussian process package!
